File: src/main/java/tv/codealong/tutorials/springboot/thenewboston/utils/db/java/AbstractServiceWithTransactionImpl.java
package tv.codealong.tutorials.springboot.thenewboston.utils.db.java;


import org.springframework.transaction.PlatformTransactionManager;
import org.springframework.transaction.TransactionDefinition;
import org.springframework.transaction.TransactionStatus;
import org.springframework.transaction.support.TransactionCallback;
import org.springframework.transaction.support.TransactionCallbackWithoutResult;
import org.springframework.transaction.support.TransactionTemplate;
import org.springframework.util.Assert;

import java.util.Set;
import java.util.function.Consumer;
import java.util.function.Function;

/**
 * Абстрактный сервис для безопасной работы с транзакциями в Spring.
 * Запрещает выполнение в "опасных" потоках (например, ForkJoinPool).
 */
public abstract class AbstractServiceWithTransactionImpl {

    private final PlatformTransactionManager transactionManager;

    // Запрещённые имена потоков для транзакций
    private static final Set<String> ILLEGAL_THREAD_NAMES = Set.of("ForkJoinPool", "coroutine");

    protected AbstractServiceWithTransactionImpl(PlatformTransactionManager transactionManager) {
        this.transactionManager = transactionManager;
    }

    /**
     * Возвращает TransactionTemplate с заданным уровнем распространения транзакции.
     */
    protected TransactionTemplate transactionTemplate(int propagationLevel) {
        TransactionTemplate tt = new TransactionTemplate(transactionManager);
        tt.setPropagationBehavior(propagationLevel);
        return tt;
    }

    /**
     * Хук: выполняется перед транзакцией. Можно переопределить.
     *
     * @return Runnable, который будет выполнен, или null.
     */
    protected Runnable beforeTransaction() {
        return null;
    }

    /**
     * Хук: выполняется после транзакции. Можно переопределить.
     *
     * @return Runnable, который будет выполнен, или null.
     */
    protected Runnable afterTransaction() {
        return null;
    }

    /**
     * Проверяет, что текущий поток разрешён для транзакций.
     */
    protected void assertTransactionAllowed() {
        String threadName = Thread.currentThread().getName();
        for (String illegalName : ILLEGAL_THREAD_NAMES) {
            Assert.doesNotContain(threadName, illegalName,
                    () -> "Illegal threadName=" + threadName + " to run transaction");
        }
    }

    /**
     * Выполняет блок кода в отдельной транзакции (REQUIRES_NEW), даже если уже есть активная.
     * Поддерживает хуки before/after.
     */
    protected void doInSeparatedTransactionWithoutResult(Consumer<TransactionStatus> block) {
        assertTransactionAllowed();
        TransactionTemplate outerTt = transactionTemplate(TransactionDefinition.PROPAGATION_NEVER);
        outerTt.executeWithoutResult(status -> {
            TransactionTemplate innerTt = transactionTemplate(TransactionDefinition.PROPAGATION_REQUIRES_NEW);
            innerTt.executeWithoutResult(innerStatus -> {
                Runnable before = beforeTransaction();
                if (before != null) before.run();
                try {
                    block.accept(innerStatus);
                } finally {
                    Runnable after = afterTransaction();
                    if (after != null) after.run();
                }
            });
        });
    }

    /**
     * Выполняет callback в отдельной транзакции (REQUIRES_NEW).
     */
    protected <T> T doInSeparatedTransaction(TransactionCallback<T> callback) {
        assertTransactionAllowed();
        TransactionTemplate outerTt = transactionTemplate(TransactionDefinition.PROPAGATION_NEVER);
        return outerTt.execute(status -> {
            TransactionTemplate innerTt = transactionTemplate(TransactionDefinition.PROPAGATION_REQUIRES_NEW);
            return doInTransaction(innerTt, callback);
        });
    }

    /**
     * Выполняет блок в обязательной внешней транзакции (MANDATORY).
     */
    protected void doInOuterMandatoryTransactionWithoutResult(Consumer<TransactionStatus> block) {
        assertTransactionAllowed();
        TransactionTemplate tt = transactionTemplate(TransactionDefinition.PROPAGATION_MANDATORY);
        tt.executeWithoutResult(status -> {
            Runnable before = beforeTransaction();
            if (before != null) before.run();
            try {
                block.accept(status);
            } finally {
                Runnable after = afterTransaction();
                if (after != null) after.run();
            }
        });
    }

    /**
     * Выполняет callback в обязательной внешней транзакции (MANDATORY).
     */
    protected <T> T doInOuterMandatoryTransaction(TransactionCallback<T> callback) {
        assertTransactionAllowed();
        return doInTransaction(transactionTemplate(TransactionDefinition.PROPAGATION_MANDATORY), callback);
    }

    /**
     * Вспомогательный метод: выполняет callback в транзакции с хуками.
     */
    private <T> T doInTransaction(TransactionTemplate tt, TransactionCallback<T> callback) {
        Runnable before = beforeTransaction();
        if (before != null) before.run();
        try {
            return tt.execute(callback);
        } finally {
            Runnable after = afterTransaction();
            if (after != null) after.run();
        }
    }
}

File: src/main/java/tv/codealong/tutorials/springboot/thenewboston/utils/db/AbstractServiceWithTransactionImpl.kt
package tv.codealong.tutorials.springboot.thenewboston.utils.db

import org.springframework.transaction.PlatformTransactionManager
import org.springframework.transaction.TransactionDefinition
import org.springframework.transaction.TransactionStatus
import org.springframework.transaction.support.TransactionCallback
import org.springframework.transaction.support.TransactionCallbackWithoutResult
import org.springframework.transaction.support.TransactionTemplate
import org.springframework.util.Assert
import java.util.function.Consumer

fun PlatformTransactionManager.withPropagation(levelPropagation: Int): TransactionTemplate = run {
    val tt = TransactionTemplate(this)
    tt.propagationBehavior = levelPropagation
    tt
}

/**
 * Абстрактный сервис для безопасной работы с транзакциями в Spring.
 * Запрещает выполнение в "опасных" потоках (например, ForkJoinPool).
 */
abstract class AbstractServiceWithTransactionImpl(
    private val transactionManager: PlatformTransactionManager
) {

    companion object {
        val illegalThreadNames = setOf("ForkJoinPool", "coroutine")
    }

    protected val transactionTemplate: (Int) -> TransactionTemplate = { levelPropagation: Int ->
        transactionManager.withPropagation(levelPropagation)
    }

    protected open fun beforeTransaction(): Runnable? {
        return null
    }

    protected open fun afterTransaction(): Runnable? {
        return null
    }

    protected fun assertTransactionAllowed() {
        val threadName = Thread.currentThread().name
        illegalThreadNames.forEach {
            Assert.doesNotContain(threadName, it) {
                "Illegal threadName=$threadName to run transaction"
            }
        }
    }

    protected open fun doInSeparatedTransactionWithoutResult(block: Consumer<TransactionStatus?>) {
        assertTransactionAllowed()
        val tt = transactionTemplate(TransactionDefinition.PROPAGATION_NEVER)
        tt.execute(object : TransactionCallbackWithoutResult() {
            override fun doInTransactionWithoutResult(transactionStatus: TransactionStatus) {
                tt.propagationBehavior = TransactionDefinition.PROPAGATION_REQUIRES_NEW
                tt.execute(object : TransactionCallbackWithoutResult() {
                    override fun doInTransactionWithoutResult(status: TransactionStatus) {
                        try {
                            beforeTransaction()?.run()
                            block.accept(status)
                        } finally {
                            afterTransaction()?.run()
                        }
                    }
                })
            }
        })
    }

    protected open fun <T> doInSeparatedTransaction(callback: TransactionCallback<T>): T? {
        assertTransactionAllowed()
        return transactionTemplate(TransactionDefinition.PROPAGATION_NEVER)
            .let { tt ->
                tt.execute {
                    tt.propagationBehavior = TransactionDefinition.PROPAGATION_REQUIRES_NEW
                    doInTransaction(tt, callback)
                }
            }
    }

    protected open fun doInOuterMandatoryTransactionWithoutResult(block: Consumer<TransactionStatus?>) {
        assertTransactionAllowed()
        val tt = transactionTemplate(TransactionDefinition.PROPAGATION_MANDATORY)
        tt.execute(object : TransactionCallbackWithoutResult() {
            override fun doInTransactionWithoutResult(transactionStatus: TransactionStatus) {
                beforeTransaction()?.run()
                try {
                    block.accept(transactionStatus)
                } finally {
                    afterTransaction()?.run()
                }
            }
        })
    }


    protected open fun <T> doInOuterMandatoryTransaction(callback: TransactionCallback<T>): T? {
        assertTransactionAllowed()
        return doInTransaction(transactionTemplate(TransactionDefinition.PROPAGATION_MANDATORY), callback)
    }

    private fun <T> doInTransaction(
        tt: TransactionTemplate,
        callback: TransactionCallback<T>
    ): T? {
        beforeTransaction()
        return try {
            tt.execute(callback)
        } finally {
            afterTransaction()
        }
    }
}